plugins/inputs/solr/solr.go

```diff
@@ -0,0 +1,395 @@
+
+		if err != nil {
+			return err
+		}
+		s.client = client
+	}
+
+	errChan := errchan.New(len(s.Servers) * len(s.Cores))
+	var wg sync.WaitGroup
+
+	for _, serv := range s.Servers {
+		for _, core := range s.Cores {
+			wg.Add(1)
+			go func(serv string, core string, acc telegraf.Accumulator) {
+				defer wg.Done()
+				if err := s.gatherCoreStats(fmt.Sprintf("%s/solr/%s"+mbeansPath, serv, core), core, acc); err != nil {
+					errChan.C <- err
+					return
+				}
+			}(serv, core, acc)
+		}
+	}
+	wg.Wait()
+	return errChan.Error()
+}
+
+func (s *Solr) createHTTPClient() (*http.Client, error) {
+	tr := &http.Transport{
+		ResponseHeaderTimeout: s.HTTPTimeout.Duration,
+	}
+	client := &http.Client{
+		Transport: tr,
+		Timeout:   s.HTTPTimeout.Duration,
+	}
+
+	return client, nil
+}
+
+func gatherCoreMetrics(mbeansJson json.RawMessage, core, category string, acc telegraf.Accumulator) error {
+	var coreMetrics map[string]Core
+
+	measurementTime := time.Now()
+	if err := json.Unmarshal(mbeansJson, &coreMetrics); err != nil {
+		return err
+	}
+	for name, metrics := range coreMetrics {
+		if strings.Contains(name, "@") {
+			continue
+		}
+		coreFields := map[string]interface{}{
+			"class_name":   metrics.Class,
+			"deleted_docs": metrics.Stats.DeletedDocs,
+			"max_docs":     metrics.Stats.MaxDoc,
+			"num_docs":     metrics.Stats.NumDocs,
+		}
+		acc.AddFields(
+			"solr_mbean_metrics",
+			coreFields,
+			map[string]string{
+				"core":    core,
+				"type":    strings.ToLower(category),
+				"handler": name},
+			measurementTime,
+		)
+	}
+	return nil
+}
+
+func gatherQueryHandlerMetrics(mbeansJson json.RawMessage, core, category string, acc telegraf.Accumulator) error {
+	var coreMetrics map[string]QueryHandler
+
+	measurementTime := time.Now()
+	if err := json.Unmarshal(mbeansJson, &coreMetrics); err != nil {
+		return err
+	}
+	for name, metrics := range coreMetrics {
+		coreFields := map[string]interface{}{
+			"class_name":                 metrics.Class,
+			"15min_rate_reqs_per_second": metrics.Stats.One5minRateReqsPerSecond,
+			"5min_rate_reqs_per_second":  metrics.Stats.FiveMinRateReqsPerSecond,
+			"75th_pc_request_time":       metrics.Stats.Seven5thPcRequestTime,
+			"95th_pc_request_time":       metrics.Stats.Nine5thPcRequestTime,
+			"999th_pc_request_time":      metrics.Stats.Nine99thPcRequestTime,
+			"99th_pc_request_time":       metrics.Stats.Nine9thPcRequestTime,
+			"avg_requests_per_second":    metrics.Stats.AvgRequestsPerSecond,
+			"avg_time_per_request":       metrics.Stats.AvgTimePerRequest,
+			"errors":                     metrics.Stats.Errors,
+			"handler_start":              metrics.Stats.HandlerStart,
+			"median_request_time":        metrics.Stats.MedianRequestTime,
+			"requests":                   metrics.Stats.Requests,
+			"timeouts":                   metrics.Stats.Timeouts,
+			"total_time":                 metrics.Stats.TotalTime,
+		}
+		acc.AddFields(
+			"solr_mbean_metrics",
```

Comment:
What about just calling it `solr` instead of `solr_mbean_metrics`. Also the same name is used in 4 different places. Should these not either have 4 different names, or be merged into one so that users don't have to deal with manually merging the 4 sets of metrics together.